File: 04-polimorfismo/python/dibujable.py
"""
dibujable.py - LA "INTERFAZ"

Python NO tiene la palabra "interface". El equivalente es otra clase
abstracta que solo declara metodos, sin datos ni constructor.

Se usa poniendola como clase base extra:
    class Circulo(Figura, Dibujable):
que es el equivalente de:
    class Circulo extends Figura implements Dibujable
"""

from abc import ABC, abstractmethod


class Dibujable(ABC):

    @abstractmethod
    def dibujar(self):
        """Contrato: quien herede esto, sabe dibujarse."""

File: 04-polimorfismo/java/Dibujable.java
/*
 * Dibujable.java - INTERFAZ
 *
 * Un contrato puro: "quien firme esto, sabe dibujar".
 * No tiene atributos ni constructor: solo la lista de lo que hay que saber hacer.
 *
 * Una clase puede heredar de UNA sola clase, pero implementar MUCHAS interfaces.
 * Por eso la interfaz vive en su propio archivo, aparte de la jerarquia.
 */
public interface Dibujable {
    void dibujar();          // en una interfaz todo es public y abstract
}
